tacacs_plus/async_client.py

```diff
@@ -0,0 +1,399 @@
+from hashlib import md5
+import sys
+import asyncio
+
+
+from .flags import (
+    TAC_PLUS_MAJOR_VER,
+    TAC_PLUS_MINOR_VER,
+    TAC_PLUS_PRIV_LVL_MIN,
+    TAC_PLUS_AUTHEN_TYPE_CHAP,
+    TAC_PLUS_MINOR_VER_ONE,
+    TAC_PLUS_AUTHEN_TYPE_PAP,
+    TAC_PLUS_AUTHEN,
+    TAC_PLUS_AUTHEN_TYPE_ASCII,
+    TAC_PLUS_AUTHEN_METH_TACACSPLUS,
+    TAC_PLUS_AUTHOR,
+    TAC_PLUS_AUTHOR_STATUS_FAIL,
+    TAC_PLUS_PRIV_LVL_MAX,
+    TAC_PLUS_ACCT,
+    TAC_PLUS_VIRTUAL_REM_ADDR,
+    TAC_PLUS_VIRTUAL_PORT,
+    TAC_PLUS_AUTHEN_STATUS_FAIL,
+    TAC_PLUS_CONTINUE_FLAG_ABORT,
+)
+from .packet import TACACSPacket, TACACSHeader
+from .authentication import (
+    TACACSAuthenticationStart,
+    TACACSAuthenticationContinue,
+    TACACSAuthenticationReply,
+)
+from .authorization import TACACSAuthorizationStart, TACACSAuthorizationReply
+from .accounting import TACACSAccountingStart, TACACSAccountingReply
+
+logger = logging.getLogger(__name__)
+
+if sys.version_info < (3, 7):
+    raise Exception(
+        'Async TACACS client is not compatible with Python version %s' % sys.version
+    )
+
+
+class TACACSClient(object):
+    """
+    A TACACS+ authentication client.
+    https://datatracker.ietf.org/doc/draft-ietf-opsawg-tacacs
+
+    An open source TACACS+ server daemon is available at
+    http://www.shrubbery.net/tac_plus/
+    """
+
+    def __init__(
+        self,
+        host,
+        port,
+        secret,
+        timeout=10,
+        session_id=None,
+        family=socket.AF_INET,
+        version_max=TAC_PLUS_MAJOR_VER,
+        version_min=TAC_PLUS_MINOR_VER,
+    ):
+        """
+        :param host:        hostname of the TACACS+ server
+        :param port:        port of the TACACS+ server, generally 49
+        :param secret:      the secret key used to obfuscate packet bodies; can
+                            be None to disable packet body obfuscation
+        :param session_id:  a unique 32-bit int representing the session; if
+                            left empty, one will be auto-generated
+        :param version_max: TACACS+ major version number, 12
+        :param version_min: TACACS+ minor version number, 0 or 1
+        """
+        self.host = host
+        self.port = port
+        self.secret = secret
+        self.timeout = timeout
+        self.version_max = version_max
+        self.version_min = version_min
+        self.family = family
+
+        # session_id is an unsigned 32-bit int; unless it's provided, randomize
+        self.session_id = session_id or random.randint(1, 2**32 - 1)
+
+    @property
+    def version(self):
+        return (self.version_max * 0x10) + self.version_min
+
+    @contextlib.asynccontextmanager
+    async def flow_control(self, reader=None, writer=None):
+        # we do not need to create new reader/writer instances
+        # if both are provided
+        existing = bool(reader and writer)
+        if existing:
+            yielded_reader, yielded_writer = reader, writer
```

Comment:
I am not sure I understand the need to pass the `reader` and `writer` to all methods. Is it to allow easier testing? 

Since we have an object here, can't they become attribute of the object that will be lazily instantiated inside `flow_control` if they do not exist?

Reply:
> > > Is it to allow easier testing?
Primarily, yes. Secondly, in case we have hundreds of requests, we can use already existing connection and save couple of seconds (moments? stress testing needed) on connection openings/closings.
To be more consistent with sync client, we can remove reader/writer params from AAA methods and create new instances on each request with no change in logic. But this params are last and optional, so not big deal.

> > > can't they become attribute of the object
We cannot move reader/writer to object attributes, as we want to properly close connections/sockets, and we want to avoid race condition. Example:
We have two simultaneous requests -- name them (1) and (2).
(1) Client creates connection and instantiates reader/writer.
(1) Client sends packet.
(2) Client sends packet. No new connection here, as reader/writer already exist.
(1) Client receive response and close connection.
(2) ??? and we have problem here.

Reply:
> To be more consistent with sync client, we can remove reader/writer params from AAA methods and create new instances on each request with no change in logic. But this params are last and optional, so not big deal.

I don't think creating new connections every time we run an operation is the right thing, it makes sense to have long lived `reader` and `writer`. I am just wondering if we can find a good way to keep the AAA methods with the same type of parameters and handle connections, which are at the end our state, internally.

In other asyncio client I have worked on, a pattern I found pretty neat is to have an async `connect()` method that creates the `reader` / `writer` and an async `disconnect()` that have to be called to properly close everything when done. 

Something along these lines:

```python
class TACACSClient(object):
    def __init__(self, ...):
        ...
        self._lock = asyncio.Lock()
        self._sock = None
        self._reader = None
        self._writer = None

    async def connect(self):
        ...
        async with self._lock:
            self._sock = socket.socket(self.family, socket.SOCK_STREAM)
            self._sock.settimeout(self.timeout)
            self._sock.connect(conn)
            self._reader, self._writer = await asyncio.open_connection(sock=sock)

    async def disconnect(self):
        async with self._lock:
            self._writer.close()
            await self.writer.wait_closed()
            self._sock.close()
            self._reader = None
            self._writer = None
            self._sock = None
    
    async def send(self, body, req_type, seq_no=1):
        if not self.reader or not self.writer:
            raise socket.error('async socket disconnected, the `connect()` was not awaited')
        ...
        self._writer.write(bytes(packet))
        await self._writer.drain()
        ...  
```

Reply:
After some investigation:

We actually need new connection on each request.

If we try to reuse connection:
 - In case of simultaneous requests: we got runtime error `read() called while another coroutine is already waiting for incoming data`, as multiple `reader.read` methods are called on same reader.
 - In case of one after another requests: server sends empty responses on all requests after first in same connection, at least my version of TACACS+ server behaves like this.

But it works well, if no reader/writer parameters were passed to AAA methods.

Solution:
Remove reader/writer parameters from AAA methods.
Think about tests.

Please accept my apologies, and i need couple of days to fix this.

Reply:
Please do not apologise and thanks for the dive deep. 

I think we can probably keep the socket as object attribute and create a new connection per request. 

Testing can be achieved with a simple pytest fixture:

```python

```

Reply:
I still couldnt think of possibility to (1) keep socket as attribute and (2) properly connect/close socket.

`asyncio.StreamReader` and `asyncio.StreamWriter` are both binded to socket, thus for different requests we need different socket instances.

Previously suggested solution is implemented in fa4527cbb91dac1d538984381aa51d82b1015e20 (thanks for monkeypatch example)

I am not sure about, and i have no chances to test `asyncio.open_connection` with IPv6 addresses. If it properly works -- we can completely get rid of socket object itself, and pass host-port directly to `open_connection`.

Reply:
Sorry for being this picky.. I am just trying to be sure we are doing the best thing and unfortunately I don't have a testing environment anymore.

I read the documentation of asyncio.open_connection and it seems like the initial idea of keeping the reader and writer as object attributes should work, but you tested and it does not :(

Let's merge the PR as it is, I think it is pretty good and I'll try to do some tests to understand if we can improve performances. Sounds good?

Reply:
Sounds awesome!